store/store.go

```diff
@@ -0,0 +1,415 @@
+	s.metrics.observeRemove(ctx, time.Since(tNow), err != nil)
+	return err
+}
+
+func (s *Store) remove(ctx context.Context, height uint64) error {
+	f, err := s.getByHeight(height)
+	if err != nil {
+		// short circuit if file not exists
+		if errors.Is(err, ErrNotFound) {
+			return nil
+		}
+		return fmt.Errorf("getting by height: %w", err)
+	}
+
+	hash, err := f.DataRoot(ctx)
+	if err != nil {
+		return fmt.Errorf("getting data hash: %w", err)
+	}
+	// close file to release the reference in the cache
+	if err = f.Close(); err != nil {
+		return fmt.Errorf("closing file on removal: %w", err)
+	}
+
+	// lock by datahash to prevent concurrent access to the same underlying file
+	// by GetByHash
+	dlock := s.stripLock.byDatahash(hash)
+	dlock.Lock()
+	defer dlock.Unlock()
+
+	if err = s.cache.Remove(height); err != nil {
+		return fmt.Errorf("removing from cache: %w", err)
+	}
+
+	// remove hard link by height
+	heightPath := s.basepath + heightsPath + fmt.Sprintf("%d", height)
+	if err = os.Remove(heightPath); err != nil {
+		return fmt.Errorf("removing by height: %w", err)
+	}
+
+	// remove file if not empty root
+	if !hash.IsEmptyRoot() {
+		hashPath := s.basepath + blocksPath + hash.String()
+		err = os.Remove(hashPath)
+		if err != nil {
+			return fmt.Errorf("removing by hash: %w", err)
+		}
+	}
+	return nil
+}
+
+func (s *Store) openFile(path string) (eds.AccessorStreamer, error) {
+	f, err := file.OpenQ1Q4File(path)
+	if err == nil {
+		return wrappedFile(f), nil
+	}
+	if os.IsNotExist(err) {
+		return nil, ErrNotFound
+	}
+	if errors.Is(err, file.ErrFileIsEmpty) {
+		return emptyAccessor, nil
+	}
+	return nil, fmt.Errorf("opening file: %w", err)
+}
+
+func fileLoader(f eds.AccessorStreamer) cache.OpenAccessorFn {
+	return func(context.Context) (eds.AccessorStreamer, error) {
+		return wrappedFile(f), nil
+	}
+}
+
+func wrappedFile(f eds.AccessorStreamer) eds.AccessorStreamer {
+	withCache := eds.WithProofsCache(f)
+	closedOnce := eds.WithClosedOnce(withCache)
+	sanityChecked := eds.WithValidation(closedOnce)
+	accessorStreamer := eds.AccessorAndStreamer(sanityChecked, closedOnce)
+	return accessorStreamer
+}
+
+func ensureFolder(path string) error {
+	info, err := os.Stat(path)
+	if os.IsNotExist(err) {
+		err = os.Mkdir(path, defaultDirPerm)
+		if err != nil {
+			return fmt.Errorf("creating blocks dir: %w", err)
+		}
+		return nil
+	}
+	if err != nil {
+		return fmt.Errorf("checking dir: %w", err)
+	}
+	if !info.IsDir() {
+		return errors.New("expected dir, got a file")
+	}
+	return nil
+}
+
+func pathExists(path string) (bool, error) {
+	_, err := os.Stat(path)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return false, nil
+		}
+		return false, err
+	}
+	return true, nil
+}
+
+func (s *Store) addEmptyHeight(height uint64) error {
+	// short circuit if link exists
+	has, err := s.hasByHeight(height)
+	if err != nil {
+		return err
+	}
+	if has {
+		return nil
+	}
+	return s.createHeightLink(share.EmptyRoot().Hash(), height)
```

Comment:
We actually don't need to do `hasByHeight` here. We can check inside `s.createHeightLink` if the returned error form `os.Link` is `os.IsExist` and if so, return success.